### ISSUE TITLE
Performance of renderItem method (Question)

### ISSUE THREAD

**Issue by tststs** [atom/autocomplete-plus#642]
Hey there,

since i was watching the timeline to find performance issues regarding `atom-ternjs` i saw this massive amount of parse HTML stacks produced by the `renderItem` method. So for 200 suggestions it looks like this:

<img width="1081" alt="bildschirmfoto 2015-12-17 um 21 50 29" src="https://cloud.githubusercontent.com/assets/3591665/11881981/46b9fec4-a50a-11e5-8eae-491612945b9f.png">

This happens quite often while developing with three.js or in bigger projects. Every time i type `THREE.` (e.g.) i'll render 200 suggestions. It feels like forever for the suggestion list to pop up.

I modified the code a bit to first generate a string that is representing the complete template of the list item. At the end of the `renderItem` method i assign `li.innerHTML` only once with the generated string.

The result looks like this:

<img width="1483" alt="bildschirmfoto 2015-12-17 um 21 52 10" src="https://cloud.githubusercontent.com/assets/3591665/11882167/4cea1b20-a50b-11e5-998a-aca749763c6f.png">

So basically my question is if you have any concerns not to reuse all of the child nodes in the cached li node in the `nodePool` and instead just reuse the `li element`. If there a no concerns i could open a PR for this.

Thanks and Greetings.
tststs

**Comment by jeancroy** [comment]
> So for 200 suggestions it looks like this

Do we want ac+ to show every suggestions ? For example, fuzzy finder will only show top N items, with N ~15

**Comment by tststs** [comment]
As to speak for myself: yes.

E.g. in some cases i need all suggestions just to scroll through and find methods or properties i can't remember or getting input of methods and properties i haven't used before.

but this is just my point of view.

Another possible solution for this is to maybe lazy-load them.

Greetings.

**Comment by Allardvm** [comment]
While looking for some low-hanging fruit to increase Atom's (perceived) performance, this also caught my attention. The `SuggestionListElement.renderItem` frequently takes several frames (>> 16.67ms) to complete because of the sheer amount of suggestions it has to render.

I agree with @tststs that some form of lazy-loading is desirable, even if some of the performance can also be improved in other ways. Large projects could have 100s of suggestions, which will always be quite taxing, even though most suggestions will not (yet) be visible.
